pinot-core/src/main/java/org/apache/pinot/core/realtime/impl/dictionary/OffHeapMutableBytesStore.java

```diff
@@ -0,0 +1,195 @@
+ * previous and current value's end offset from the same buffer.
+ */
+public class OffHeapMutableBytesStore implements Closeable {
+  private static final byte[] EMPTY_BYTES = new byte[0];
+
+  // MAX_NUM_BUFFERS = 8_192
+  private static final int MAX_NUM_BUFFERS = 1 << 13;
+
+  // OFFSET_BUFFER_SIZE = 32_768 + 4
+  private static final int OFFSET_BUFFER_SHIFT_OFFSET = 13;
+  private static final int OFFSET_BUFFER_SIZE = ((1 << OFFSET_BUFFER_SHIFT_OFFSET) + 1) << 2;
+  private static final int OFFSET_BUFFER_MASK = 0xFFFFFFFF >>> (Integer.SIZE - OFFSET_BUFFER_SHIFT_OFFSET);
+
+  // VALUE_BUFFER_SIZE = 1_048_576
+  private static final int VALUE_BUFFER_SIFT_OFFSET = 20;
+  private static final int VALUE_BUFFER_SIZE = 1 << VALUE_BUFFER_SIFT_OFFSET;
+  private static final int VALUE_BUFFER_MASK = 0xFFFFFFFF >>> (Integer.SIZE - VALUE_BUFFER_SIFT_OFFSET);
+
+  private final PinotDataBuffer[] _offsetBuffers = new PinotDataBuffer[MAX_NUM_BUFFERS];
+  private final PinotDataBuffer[] _valueBuffers = new PinotDataBuffer[MAX_NUM_BUFFERS];
+  private final PinotDataBufferMemoryManager _memoryManager;
+  private final String _allocationContext;
+
+  private int _numValues;
+  private int _valueStartOffset;
+  private int _totalBufferSize;
+
+  public OffHeapMutableBytesStore(PinotDataBufferMemoryManager memoryManager, String allocationContext) {
+    _memoryManager = memoryManager;
+    _allocationContext = allocationContext;
+  }
+
+  public int add(byte[] value) {
+    int offsetBufferIndex = _numValues >>> OFFSET_BUFFER_SHIFT_OFFSET;
+    int offsetIndex = _numValues & OFFSET_BUFFER_MASK;
+
+    PinotDataBuffer offsetBuffer;
+    // If this is the first entry in the offset buffer, allocate a new buffer and store the start offset (end offset of
+    // the previous value)
+    if (offsetIndex == 0) {
+      offsetBuffer = _memoryManager.allocate(OFFSET_BUFFER_SIZE, _allocationContext);
+      offsetBuffer.putInt(0, _valueStartOffset);
+      _offsetBuffers[offsetBufferIndex] = offsetBuffer;
+      _totalBufferSize += OFFSET_BUFFER_SIZE;
+    } else {
+      offsetBuffer = _offsetBuffers[offsetBufferIndex];
+    }
+
+    int valueLength = value.length;
+    if (valueLength == 0) {
+      offsetBuffer.putInt((offsetIndex + 1) << 2, _valueStartOffset);
+      return _numValues++;
+    }
+
+    int valueBufferIndex = (_valueStartOffset + valueLength - 1) >>> VALUE_BUFFER_SIFT_OFFSET;
+    PinotDataBuffer valueBuffer;
+    // If the current value buffer does not have enough space, allocate a new buffer to store the value
+    if ((_valueStartOffset - 1) >>> VALUE_BUFFER_SIFT_OFFSET != valueBufferIndex) {
+      valueBuffer = _memoryManager.allocate(VALUE_BUFFER_SIZE, _allocationContext);
+      _valueBuffers[valueBufferIndex] = valueBuffer;
+      _totalBufferSize += VALUE_BUFFER_SIZE;
+      _valueStartOffset = valueBufferIndex << VALUE_BUFFER_SIFT_OFFSET;
```

Comment:
Doesn't this statement limit MAX_NUM_BUFFERS to (Integer.SIZE - VALUE_BUFFER_SHIFT_OFFSET)?

Reply:
No, these 2 are independent.
We use the shift operation to quickly get the startOffset because it is the first value in the current value buffer.

Reply:
valueBufferIndex should fit in (Integer.SIZE - VALUE_BUFFER_SHIFT_OFFSET - 1) bits. Therefore, MAX_NUM_BUFFERS should be at most (2 ** (Integer.SIZE - VALUE_BUFFER_SHIFT_OFFSET - 1)).

If valueBufferIndex does not fit in these bits, then we have a problem with overflow in the expression (valueBufferIndex << VALUE_BUFFER_SHIFT_OFFSET)

Right?

Reply:
Oh I see what you mean. Very good question. Yes, we should not store more than 2GB data in this bytes store, or we might trigger the int overflow for the global offset. This limitation is listed in the comments

Reply:
So, if we can add a pre-conditions check in the constructor that will be useful.  As written, it seems to imply that the static definitions can be set to whatever and things will work.

Reply:
The limitation comes from the int type for offset, but not these constants. Meaning it still works for any buffer size/number of buffers, but we can only hold 2GB data.